lib/global-styles.php

```diff
@@ -253,64 +267,101 @@ function gutenberg_experimental_global_styles_get_theme() {
-			'selector' => 'h1',
-			'supports' => array( 'line-height', 'font-size', 'color' ),
-		),
-		'core/heading/h2' => array(
-			'selector' => 'h2',
-			'supports' => array( 'line-height', 'font-size', 'color' ),
-		),
-		'core/heading/h3' => array(
-			'selector' => 'h3',
-			'supports' => array( 'line-height', 'font-size', 'color' ),
-		),
-		'core/heading/h4' => array(
-			'selector' => 'h4',
-			'supports' => array( 'line-height', 'font-size', 'color' ),
-		),
-		'core/heading/h5' => array(
-			'selector' => 'h5',
-			'supports' => array( 'line-height', 'font-size', 'color' ),
-		),
-		'core/heading/h6' => array(
-			'selector' => 'h6',
-			'supports' => array( 'line-height', 'font-size', 'color' ),
-		),
-		'core/columns'    => array(
-			'selector' => '.wp-block-columns',
-			'supports' => array( 'color' ),
-		),
-		'core/group'      => array(
-			'selector' => '.wp-block-group',
-			'supports' => array( 'color' ),
-		),
-		'core/media-text' => array(
-			'selector' => '.wp-block-media-text',
-			'supports' => array( 'color' ),
+			'supports' => array( 'background-color' ),
 		),
 	);
 
+	$registry = WP_Block_Type_Registry::get_instance();
+	foreach ( $registry->get_all_registered() as $block_name => $block_type ) {
+		if ( ! is_array( $block_type->supports ) ) {
+			continue;
+		}
+
+		$supports = gutenberg_experimental_global_styles_get_supported_styles( $block_type->supports );
+		if ( empty( $supports ) ) {
+			continue;
+		}
+
+		/*
+		 * Assign the selector for the block.
+		 *
+		 * Some blocks can declare multiple selectors:
+		 *
+		 * - core/heading represents the H1-H6 HTML elements
+		 * - core/list represents the UL and OL HTML elements
+		 * - core/group is meant to represent DIV and other HTML elements
+		 *
+		 * Some other blocks don't provide a selector,
+		 * so we generate a class for them based on their name:
+		 *
+		 * - 'core/group' => '.wp-block-group'
+		 * - 'my-custom-library/block-name' => '.wp-block-my-custom-library-block-name'
+		 *
+		 * Note that, for core blocks, we don't add the `core/` prefix to its class name.
+		 * This is for historical reasons, as they come with a class without that infix.
+		 *
+		 */
+		if (
+			isset( $block_type->supports['__experimentalSelector'] ) &&
+			is_string( $block_type->supports['__experimentalSelector'] )
+		) {
+			$block_data[ $block_name ] = array(
+				'selector' => $block_type->supports['__experimentalSelector'],
+				'supports' => $supports,
+			);
+		} elseif (
+			isset( $block_type->supports['__experimentalSelector'] ) &&
+			is_array( $block_type->supports['__experimentalSelector'] )
+		) {
+			foreach ( $block_type->supports['__experimentalSelector'] as $key => $selector ) {
+				$block_data[ $key ] = array(
+					'selector' => $selector,
+					'supports' => $supports,
+				);
+			}
+		} else {
+			$block_data[ $block_name ] = array(
+				'selector' => '.wp-block-' . preg_replace( '/\//', '-', preg_replace( '/core\//', '', $block_name ) ),
```

Comment:
We are using a regex replace with preg_replace which is more costly that a simple str_replace. I think in this case we are not taking advantage of a regex so str_replace should be enough.

Reply:
Changed it to `str_replace` at https://github.com/WordPress/gutenberg/pull/22698/commits/b2a77df59d3fda32f84a8072e6ec09c660f4326f